### ISSUE TITLE
Clarification: non-existent subproject through dependency-fallback

### ISSUE THREAD

**Issue by kusma** [mesonbuild/meson#8710]
In mesa, we currently have a non-existent subproject as a fallback for a missing LLVM.

The reason we don't (can't?) provide an actual subproject, is that we don't think we can reasonably provide a universally correct configuration for LLVM. It's a complicated project to build, with a lot of knobs that users might want to turn.

We've had it this way for years, but when a recent driver-update bumped the LLVM requirement, I got the following confusing error:

> Subproject directory not found and llvm.wrap file not found

I'm using my distro-provided LLVM packages, so I immediately got puzzled about the mention of subprojects or wraps before I realized that it was a red herring, and the real issue was that I simply hadn't installed a recent enough version of LLVM. That error message is confusing in my case, but it's probably the right one in some cases.

I tried figuring out if we were doing something wrong, and then I encountered the following things in the docs:

1. The [`subproject()` docs](https://mesonbuild.com/Reference-manual.html#subproject) say:
    > Subprojects must always be placed inside the `subprojects` directory at the top source directory. 
2. The [`dependency()` docs](https://mesonbuild.com/Reference-manual.html#dependency) say:
   > [...] by a fallback subproject which, if needed, will be brought into the current build specification as if `subproject()` had been called. 

When combined, these two sentences suggests to me that it's an error in the build system to declare a fallback to a subproject that doesn't exist.

But, then I discovered that there's seemingly a contradiction in the docs, where the `dependency()` docs *also* say:
> Note that this means the fallback dependency may be a not-found dependency, in which case the value of the `required:` kwarg will be obeyed.

The latter makes me think that the wording of the docs should be softened up a bit to reflect that there's cases where this is legal.

You can see [this Mesa MR](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10499) for the discussion on the Mesa side.

**Comment by xclaesse** [comment]
I see mesa has `fallback : ['llvm', 'dep_llvm'],` but does not ship a llvm.wrap file? Just remove that fallback then... If what you want is to allow some people to write their own llvm.wrap in the case they really want to fallback, then you should instead rely on the `[provide]` section in the wrap file and remove the fallback from that `dependency()` call.

See https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section.

**Comment by xclaesse** [comment]
> But, then I discovered that there's seemingly a contradiction in the docs, where the `dependency()` docs _also_ say:
> 
> > Note that this means the fallback dependency may be a not-found dependency, in which case the value of the `required:` kwarg will be obeyed.
> 
> The latter makes me think that the wording of the docs should be softened up a bit to reflect that there's cases where this is legal.

What that part of the doc means is when you do `fallback : ['llvm', 'dep_llvm']` it takes the `dep_llvm` variable from llvm subproject. But that variable could itself be a not-found dependency object (unlikely, but could be). In that case if you have `required: true` it will abort with not-found dep error message, but if `required: false` then it will just return the not-found dependency object and continue.

**Comment by kusma** [comment]
OK, so I guess my interpretation of the latter quote is incorrect, and this is incorrect as-is in mesa?

**Comment by dcbaker** [comment]
We offer that so that vmware can use a binary wrap with their pre-built llvm on windows. If we could ever bump the version requirements of meson we could remove the explicit fallback, right @xclaesse?

**Comment by tristan957** [comment]
Anything actionable on the Meson side here? Seems like Mesa could do:

```meson
kwargs = {}
if meson.version().version_compate('<=0.52')
  kwargs += {'fallback': ['llvm', 'dep_llvm']}
endif
dependency('llvm', kwargs: kwargs)
```
or whatever that hidden syntax is.

**Comment by kusma** [comment]
@tristan957: Well, yes... Clarifying the docs.

I'm really trying to figure out two things:
1. Is what we're doing in Mesa even legal? Most discussion so far seems to be about this
2. What are the actual rules here? The docs seems confused about this.

The actionable part would be to clarify the docs. The different parts seems contradictionary to me.

**Comment by eli-schwartz** [comment]
I think the best way to describe this is "technically not illegal".

Attempting to *use* a fallback, when a fallback cannot be analyzed as available to "try", spawns an error message stating that. But only if the dependency is required.

If the dependency is not required, we log the error but continue. If the dependency *is* required, this is probably getting assumed to be the "most specific" error message and thus used as is.

**Comment by kusma** [comment]
OK, in that case:

Is this legal for the `subproject()` function? If so, I would rephrase "Subprojects must always be placed inside [...]" to "`Subprojects should be placed inside [...]"

If that's not the case, I would rephrase "as if subproject() had been called" to "similar to if subproject() had been called".
